go/mysql/flavor_mysql.go

```diff
@@ -332,17 +332,30 @@ LEFT OUTER JOIN (
 	FROM information_schema.innodb_sys_tablespaces
 	WHERE name LIKE CONCAT(database(), '/%')
 	GROUP BY space, file_size, allocated_size, name
-) i ON i.name = CONCAT(t.table_schema, '/', t.table_name) or i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%')
+) i ON i.name = CONCAT(t.table_schema, '/', t.table_name) OR i.name LIKE CONCAT(t.table_schema, '/', t.table_name, '#p#%')
 WHERE t.table_schema = database()
-GROUP BY t.table_name, t.table_type, t.create_time, t.table_comment
-`
+GROUP BY t.table_name, t.table_type, t.create_time, t.table_comment`
 
 // TablesWithSize80 is a query to select table along with size for mysql 8.0
-const TablesWithSize80 = `SELECT t.table_name, t.table_type, unix_timestamp(t.create_time), t.table_comment, sum(i.file_size), sum(i.allocated_size) 
-		FROM information_schema.tables t, information_schema.innodb_tablespaces i 
-		WHERE t.table_schema = database() and 
-		(i.name = concat(t.table_schema,'/',t.table_name) or i.name like concat(t.table_schema,'/',t.table_name, '#p#%')) 
-		group by t.table_name, t.table_type, t.create_time, t.table_comment, i.file_size`
+//
+// We join with a subquery that materializes the data from `information_schema.innodb_sys_tablespaces`
```

Comment:
Would you mind pasting the EXPLAIN output from mysql that shows this improvement?

Reply:
Here's the explain plan for the original query:

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "158.54"
    },
    "grouping_operation": {
      "using_temporary_table": true,
      "using_filesort": false,
      "nested_loop": [
        {
          "table": {
            "table_name": "cat",
            "access_type": "index",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "name",
            "used_key_parts": [
              "name"
            ],
            "key_length": "194",
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 1,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "0.25",
              "eval_cost": "0.10",
              "prefix_cost": "0.35",
              "data_read_per_join": "224"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "sch",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY",
              "catalog_id"
            ],
            "key": "catalog_id",
            "used_key_parts": [
              "catalog_id",
              "name"
            ],
            "key_length": "202",
            "ref": [
              "mysql.cat.id",
              "const"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 1,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "0.25",
              "eval_cost": "0.10",
              "prefix_cost": "0.70",
              "data_read_per_join": "256"
            },
            "used_columns": [
              "id",
              "catalog_id",
              "name"
            ]
          }
        },
        {
          "table": {
            "table_name": "i",
            "access_type": "ALL",
            "using_join_buffer": "hash join",
            "used_columns": [
              "SPACE",
              "NAME",
              "FLAG",
              "ROW_FORMAT",
              "PAGE_SIZE",
              "ZIP_PAGE_SIZE",
              "SPACE_TYPE",
              "FS_BLOCK_SIZE",
              "FILE_SIZE",
              "ALLOCATED_SIZE",
              "AUTOEXTEND_SIZE",
              "SERVER_VERSION",
              "SPACE_VERSION",
              "ENCRYPTION",
              "STATE"
            ]
          }
        },
        {
          "table": {
            "table_name": "tbl",
            "access_type": "ref",
            "possible_keys": [
              "schema_id"
            ],
            "key": "schema_id",
            "used_key_parts": [
              "schema_id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.sch.id"
            ],
            "rows_examined_per_scan": 49,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "cost_info": {
              "read_cost": "77.78",
              "eval_cost": "9.98",
              "prefix_cost": "91.18",
              "data_read_per_join": "3M"
            },
            "used_columns": [
              "id",
              "schema_id",
              "name",
              "type",
              "collation_id",
              "comment",
              "hidden",
              "options",
              "tablespace_id",
              "created"
            ],
            "attached_condition": "(((`i`.`NAME` = concat(`mysql`.`sch`.`name`,'/',`mysql`.`tbl`.`name`)) or (`i`.`NAME` like concat(`mysql`.`sch`.`name`,'/',`mysql`.`tbl`.`name`,'#p#%'))) and (0 <> can_access_table(`mysql`.`sch`.`name`,`mysql`.`tbl`.`name`)) and (0 <> is_visible_dd_object(`mysql`.`tbl`.`hidden`)))"
          }
        },
        {
          "table": {
            "table_name": "col",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.tbl.collation_id"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "113.63",
              "data_read_per_join": "22K"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "ts",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.tbl.tablespace_id"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "136.09",
              "data_read_per_join": "700K"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "stat",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "schema_name",
              "table_name"
            ],
            "key_length": "388",
            "ref": [
              "const",
              "mysql.tbl.name"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "158.54",
              "data_read_per_join": "45K"
            },
            "used_columns": [
              "schema_name",
              "table_name"
            ]
          }
        }
      ]
    }
  }
}
```

Here's the explain plan for the new query:

```json
{
  "query_block": {
    "select_id": 1,
    "cost_info": {
      "query_cost": "158.54"
    },
    "grouping_operation": {
      "using_temporary_table": true,
      "using_filesort": false,
      "nested_loop": [
        {
          "table": {
            "table_name": "cat",
            "access_type": "index",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "name",
            "used_key_parts": [
              "name"
            ],
            "key_length": "194",
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 1,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "0.25",
              "eval_cost": "0.10",
              "prefix_cost": "0.35",
              "data_read_per_join": "224"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "sch",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY",
              "catalog_id"
            ],
            "key": "catalog_id",
            "used_key_parts": [
              "catalog_id",
              "name"
            ],
            "key_length": "202",
            "ref": [
              "mysql.cat.id",
              "const"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 1,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "0.25",
              "eval_cost": "0.10",
              "prefix_cost": "0.70",
              "data_read_per_join": "256"
            },
            "used_columns": [
              "id",
              "catalog_id",
              "name"
            ]
          }
        },
        {
          "table": {
            "table_name": "i",
            "access_type": "ALL",
            "rows_examined_per_scan": 2,
            "rows_produced_per_join": 2,
            "filtered": "100.00",
            "using_join_buffer": "hash join",
            "cost_info": {
              "read_cost": "2.53",
              "eval_cost": "0.20",
              "prefix_cost": "3.43",
              "data_read_per_join": "3K"
            },
            "used_columns": [
              "space",
              "file_size",
              "allocated_size",
              "name"
            ],
            "materialized_from_subquery": {
              "using_temporary_table": true,
              "dependent": false,
              "cacheable": true,
              "query_block": {
                "select_id": 2,
                "cost_info": {
                  "query_cost": "2.72"
                },
                "grouping_operation": {
                  "using_temporary_table": true,
                  "using_filesort": false,
                  "table": {
                    "table_name": "innodb_tablespaces",
                    "access_type": "ALL",
                    "used_columns": [
                      "SPACE",
                      "NAME",
                      "FLAG",
                      "ROW_FORMAT",
                      "PAGE_SIZE",
                      "ZIP_PAGE_SIZE",
                      "SPACE_TYPE",
                      "FS_BLOCK_SIZE",
                      "FILE_SIZE",
                      "ALLOCATED_SIZE",
                      "AUTOEXTEND_SIZE",
                      "SERVER_VERSION",
                      "SPACE_VERSION",
                      "ENCRYPTION",
                      "STATE"
                    ],
                    "attached_condition": "(`innodb_tablespaces`.`NAME` like <cache>(concat(database(),'/%')))"
                  }
                }
              }
            }
          }
        },
        {
          "table": {
            "table_name": "tbl",
            "access_type": "ref",
            "possible_keys": [
              "schema_id"
            ],
            "key": "schema_id",
            "used_key_parts": [
              "schema_id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.sch.id"
            ],
            "rows_examined_per_scan": 49,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "cost_info": {
              "read_cost": "77.78",
              "eval_cost": "9.98",
              "prefix_cost": "91.18",
              "data_read_per_join": "3M"
            },
            "used_columns": [
              "id",
              "schema_id",
              "name",
              "type",
              "collation_id",
              "comment",
              "hidden",
              "options",
              "tablespace_id",
              "created"
            ],
            "attached_condition": "(((`i`.`name` = concat(`mysql`.`sch`.`name`,'/',`mysql`.`tbl`.`name`)) or (`i`.`name` like concat(`mysql`.`sch`.`name`,'/',`mysql`.`tbl`.`name`,'#p#%'))) and (0 <> can_access_table(`mysql`.`sch`.`name`,`mysql`.`tbl`.`name`)) and (0 <> is_visible_dd_object(`mysql`.`tbl`.`hidden`)))"
          }
        },
        {
          "table": {
            "table_name": "col",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.tbl.collation_id"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "113.63",
              "data_read_per_join": "22K"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "ts",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "id"
            ],
            "key_length": "8",
            "ref": [
              "mysql.tbl.tablespace_id"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "136.09",
              "data_read_per_join": "700K"
            },
            "used_columns": [
              "id"
            ]
          }
        },
        {
          "table": {
            "table_name": "stat",
            "access_type": "eq_ref",
            "possible_keys": [
              "PRIMARY"
            ],
            "key": "PRIMARY",
            "used_key_parts": [
              "schema_name",
              "table_name"
            ],
            "key_length": "388",
            "ref": [
              "const",
              "mysql.tbl.name"
            ],
            "rows_examined_per_scan": 1,
            "rows_produced_per_join": 99,
            "filtered": "100.00",
            "using_index": true,
            "cost_info": {
              "read_cost": "12.47",
              "eval_cost": "9.98",
              "prefix_cost": "158.54",
              "data_read_per_join": "45K"
            },
            "used_columns": [
              "schema_name",
              "table_name"
            ]
          }
        }
      ]
    }
  }
}
```

Reply:
Here's also the `EXPLAIN ANALYZE` output for the old query:

```
EXPLAIN: -> Table scan on <temporary>  (actual time=0.002..0.017 rows=131 loops=1)
    -> Aggregate using temporary table  (actual time=525.326..525.349 rows=131 loops=1)
        -> Nested loop left join  (actual time=219.629..524.812 rows=131 loops=1)
            -> Nested loop left join  (actual time=219.616..524.617 rows=131 loops=1)
                -> Nested loop left join  (actual time=219.611..524.580 rows=131 loops=1)
                    -> Nested loop inner join  (actual time=219.599..524.504 rows=131 loops=1)
                        -> Inner hash join (no condition)  (actual time=22.868..24.513 rows=2464 loops=1)
                            -> Table scan on i  (cost=0.00 rows=0) (actual time=0.002..1.116 rows=2464 loops=1)
                                -> Fill information schema table i  (actual time=22.821..24.071 rows=2464 loops=1)
                            -> Hash
                                -> Nested loop inner join  (cost=0.70 rows=1) (actual time=0.032..0.036 rows=1 loops=1)
                                    -> Covering index scan on cat using name  (cost=0.35 rows=1) (actual time=0.018..0.021 rows=1 loops=1)
                                    -> Single-row covering index lookup on sch using catalog_id (catalog_id=mysql.cat.id, name=database())  (cost=0.35 rows=1) (actual time=0.013..0.013 rows=1 loops=1)
                        -> Filter: (((i.`NAME` = concat(mysql.sch.`name`,'/',mysql.tbl.`name`)) or (i.`NAME` like concat(mysql.sch.`name`,'/',mysql.tbl.`name`,'#p#%'))) and (0 <> can_access_table(mysql.sch.`name`,mysql.tbl.`name`)) and (0 <> is_visible_dd_object(mysql.tbl.hidden)))  (cost=40.50 rows=50) (actual time=0.197..0.203 rows=0 loops=2464)
                            -> Index lookup on tbl using schema_id (schema_id=mysql.sch.id)  (cost=40.50 rows=50) (actual time=0.002..0.152 rows=131 loops=2464)
                    -> Single-row covering index lookup on col using PRIMARY (id=mysql.tbl.collation_id)  (cost=0.13 rows=1) (actual time=0.000..0.000 rows=1 loops=131)
                -> Single-row covering index lookup on ts using PRIMARY (id=mysql.tbl.tablespace_id)  (cost=0.13 rows=1) (actual time=0.000..0.000 rows=0 loops=131)
            -> Single-row covering index lookup on stat using PRIMARY (schema_name=database(), table_name=mysql.tbl.`name`)  (cost=0.13 rows=1) (actual time=0.001..0.001 rows=0 loops=131)

1 row in set (0.53 sec)
```

And the new query:

```
EXPLAIN: -> Table scan on <temporary>  (actual time=0.001..0.012 rows=131 loops=1)
    -> Aggregate using temporary table  (actual time=55.406..55.426 rows=131 loops=1)
        -> Nested loop left join  (cost=116.66 rows=0) (actual time=25.128..54.991 rows=131 loops=1)
            -> Nested loop left join  (cost=104.19 rows=0) (actual time=25.119..54.810 rows=131 loops=1)
                -> Nested loop left join  (cost=91.72 rows=0) (actual time=25.115..54.776 rows=131 loops=1)
                    -> Nested loop inner join  (cost=79.24 rows=0) (actual time=25.108..54.712 rows=131 loops=1)
                        -> Inner hash join (no condition)  (cost=3.23 rows=0) (actual time=25.058..25.108 rows=131 loops=1)
                            -> Table scan on i  (cost=2.50..2.50 rows=0) (actual time=0.001..0.014 rows=131 loops=1)
                                -> Materialize  (cost=5.00..5.00 rows=0) (actual time=24.994..25.015 rows=131 loops=1)
                                    -> Table scan on <temporary>  (cost=2.50..2.50 rows=0) (actual time=0.001..0.010 rows=131 loops=1)
                                        -> Temporary table with deduplication  (cost=2.50..2.50 rows=0) (actual time=24.930..24.950 rows=131 loops=1)
                                            -> Filter: (innodb_tablespaces.`NAME` like <cache>(concat(database(),'/%')))  (actual time=23.968..24.800 rows=131 loops=1)
                                                -> Table scan on innodb_tablespaces  (cost=0.00 rows=0) (actual time=0.002..0.894 rows=2464 loops=1)
                                                    -> Fill information schema table innodb_tablespaces  (actual time=23.439..24.459 rows=2464 loops=1)
                            -> Hash
                                -> Nested loop inner join  (cost=0.70 rows=1) (actual time=0.047..0.051 rows=1 loops=1)
                                    -> Covering index scan on cat using name  (cost=0.35 rows=1) (actual time=0.035..0.040 rows=1 loops=1)
                                    -> Single-row covering index lookup on sch using catalog_id (catalog_id=mysql.cat.id, name=database())  (cost=0.35 rows=1) (actual time=0.011..0.011 rows=1 loops=1)
                        -> Filter: (((i.`name` = concat(mysql.sch.`name`,'/',mysql.tbl.`name`)) or (i.`name` like concat(mysql.sch.`name`,'/',mysql.tbl.`name`,'#p#%'))) and (0 <> can_access_table(mysql.sch.`name`,mysql.tbl.`name`)) and (0 <> is_visible_dd_object(mysql.tbl.hidden)))  (cost=40.50 rows=50) (actual time=0.113..0.226 rows=1 loops=131)
                            -> Index lookup on tbl using schema_id (schema_id=mysql.sch.id)  (cost=40.50 rows=50) (actual time=0.002..0.151 rows=131 loops=131)
                    -> Single-row covering index lookup on col using PRIMARY (id=mysql.tbl.collation_id)  (cost=0.13 rows=1) (actual time=0.000..0.000 rows=1 loops=131)
                -> Single-row covering index lookup on ts using PRIMARY (id=mysql.tbl.tablespace_id)  (cost=0.13 rows=1) (actual time=0.000..0.000 rows=0 loops=131)
            -> Single-row covering index lookup on stat using PRIMARY (schema_name=database(), table_name=mysql.tbl.`name`)  (cost=0.13 rows=1) (actual time=0.001..0.001 rows=0 loops=131)

1 row in set (0.06 sec)
```